src/Parallel/ElementRegistration.hpp

```diff
@@ -0,0 +1,80 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <type_traits>
+
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace db {
+template <typename DbTagList>
+class DataBox;
+}  // namespace db
+namespace Parallel {
+template <typename Metavariables>
+class GlobalCache;
+}  // namespace Parallel
+/// \endcond
+
+namespace Parallel {
+namespace detail {
+template <typename Metavariables, typename ParallelComponent,
+          typename = std::void_t<>>
+struct has_registration_list : std::false_type {};
+
+template <typename Metavariables, typename ParallelComponent>
+struct has_registration_list<
+    Metavariables, ParallelComponent,
+    std::void_t<typename Metavariables::template registration_list<
+        ParallelComponent>::type>> : std::true_type {};
+
+template <typename Metavariables, typename ParallelComponent>
+constexpr bool has_registration_list_v =
+    has_registration_list<Metavariables, ParallelComponent>::value;
+
+template <typename Metavariables, typename ParallelComponent>
+using registration_list =
+    typename Metavariables::template registration_list<ParallelComponent>::type;
+}  // namespace detail
+
+template <typename ParallelComponent, typename DbTagList,
+          typename Metavariables, typename ArrayIndex>
+void deregister_element(db::DataBox<DbTagList>& box,
+                        Parallel::GlobalCache<Metavariables>& cache,
```

Comment:
take these things as not-null pointers?

Reply:
I'm keeping as-is as we use this pattern throughout the action framework when passing the DataBox, GlobalCache, and ArrayIndex around.